### PR TITLE
Added checks to prevent adding rules with the same name

### PR DESCRIPTION
**Pull request by nicosp**
Added checks to prevent ORM rules added with the same name. Currently the rules are silently replaced which is not ideal.

Related issue: https://github.com/cakephp/cakephp/issues/17669

### CODE REVIEW

File: src/Datasource/Exception/RuleAlreadyExistsException.php [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+ * @license       https://opensource.org/licenses/mit-license.php MIT License
+ */
+namespace Cake\Datasource\Exception;
+
+use Cake\Core\Exception\CakeException;
+
+/**
+ * Exception raised when a rule with the same name already exists
+ */
+class RuleAlreadyExistsException extends CakeException
```
Comment (ADmad) [inline src/Datasource/Exception/RuleAlreadyExistsException.php]: Given how uncommon this issue is and it's very likely to be caught at development time itself, I am not sure a separate exception is needed. We could just throw `CakeException`.

Let's see what others have to say.

File: src/Datasource/RulesChecker.php [hunk 2]
```diff
@@ -404,4 +413,19 @@ protected function _addError(callable $rule, array|string|null $name = null, arr
...
+
+    /**
+     * Checks that a rule with the same name doesn't already exist
+     *
+     * @param array<\Cake\Datasource\RuleInvoker> $rules The rules array to check
+     * @param string $name The name to check
+     * @return void
+     * @throws \Cake\Datasource\Exception\RuleAlreadyExistsException
+     */
+    private function _checkName(array $rules, string $name): void
```
Comment (ADmad) [inline src/Datasource/RulesChecker.php]:
```suggestion
    protected function checkName(array $rules, string $name): void
```

File: src/Datasource/RulesChecker.php [hunk 3]
```diff
@@ -404,4 +413,19 @@ protected function _addError(callable $rule, array|string|null $name = null, arr
...
+
+    /**
+     * Checks that a rule with the same name doesn't already exist
+     *
+     * @param array<\Cake\Datasource\RuleInvoker> $rules The rules array to check
+     * @param string $name The name to check
+     * @return void
+     * @throws \Cake\Datasource\Exception\RuleAlreadyExistsException
+     */
+    protected function _checkName(array $rules, string $name): void
```
Comment (ADmad) [inline src/Datasource/RulesChecker.php]:
```suggestion
    protected function checkName(string $name, array $rules): void
```

We no longer add underscore prefix to protected properties/methods. Also having name as the 1st param feels more natural, that's the order for array_key_exists() too.

File: src/Datasource/Exception/RuleAlreadyExistsException.php (hunk 1 above)
Comment (markstory) [reply src/Datasource/Exception/RuleAlreadyExistsException.php]: I agree with @ADmad. This isn't a type that will have application level catch blocks and we won't have a specific error page either. `CakeException` should be good enough here.

**Pull request merged**